File: src/leetCode/java/Solution110.java
package leetCode.java;

//DFS O(V) O(H)
class Solution110 {
    public boolean isBalanced(TreeNode root) {
        int height = dfs(root);
        return height != -1;
    }

    public int dfs(TreeNode root){
        if(root == null) return 0;

        int lHeight = dfs(root.left);
        int rHeight = dfs(root.right);
        //height differ more than 1
        if(lHeight == -1 || rHeight == -1 || Math.abs(lHeight - rHeight) > 1)
            return -1;//表示false

        return Math.max(lHeight, rHeight) + 1;
    }
}

File: src/leetCode/c++/Solution110.cpp
//DFS O(V) O(H)
class Solution110 {
public:
    bool isBalanced(TreeNode* root) {
        return dfs(root) != -1;
    }

    int dfs(TreeNode* root){
        if(!root) return 0;

        int left_height = dfs(root->left);
        int right_height = dfs(root->right);
        //node differ in height by no more than 1.
        if(left_height == -1 || right_height == -1 || abs(left_height - right_height) > 1)
            return -1;

        return max(left_height, right_height) + 1;
    }
};